### ISSUE TITLE
[ENHANCEMENT] Add support for CriticMarkup

### ISSUE THREAD

**Issue by rsgranne** [Zettlr/Zettlr#728]
## Description

It would be great if you could add support for [CriticMarkup](http://criticmarkup.com/). It’s become widely used in Markdown circles & apps, & support for it here would add some useful functionality. 

For those unfamiliar with CriticMarkup, here’s their description from their website:

> CriticMarkup is a way for authors and editors to track changes to documents in plain text. As with Markdown, small groups of distinctive characters allow you to highlight insertions, deletions, substitutions and comments, all without the overhead of heavy, proprietary office suites.

Other Markdown editors, such as BBEdit & Sublime Text, support it, as do macOS macro programs, such as Keyboard Maestro & TextExpander. The [GitHub page](https://github.com/CriticMarkup/CriticMarkup-toolkit) has more info.

## Proposed Changes

With CriticMarkup, users have available these 5 markup additions:

* Addition {++ ++}
* Deletion {-- --}
* Substitution {~~ ~> ~~}
* Comment {>> <<}
* Highlight {== ==}{>> <<}

These turn out to be extremely useful.

## Caveats

None that I am aware of.

## Additional Information

I am thinking about switching from nvUltra to Zettlr. nvUltra currently supports CriticMarkup, & I find it immensely useful. Adding that feature to Zettlr would be great for users who desire that functionality.

**Comment by nathanlesage** [comment]
Puhhh … I had a look at this two years ago, but wasn't really happy with how they did it … I still think that a `git diff` is a much better way … opinions?

**Comment by enricllagostera** [comment]
I like the idea of the toolkit as an option for editing and reviewing, it does seem more readable and less git-dependent. I've never used it for long though. I think it does work well as a way of bypassing workflows that involve edit tracking in Word / GDocs.

**Comment by nathanlesage** [comment]
> I like the idea of the toolkit as an option for editing and reviewing, it does seem more readable and less git-dependent. I've never used it for long though. I think it does work well as a way of bypassing workflows that involve edit tracking in Word / GDocs.

Hm … actually, I just thought that converting Word annotations to Critic Markdown would actually be much easier with Critic Markup 🧐

**Comment by rsgranne** [comment]
Hey, I like diffs as much as they next guy, but as I'm trying to turn on college students to the wonders of plain text & Markdown, the vast majority of them will never use git or diff.

**Comment by nathanlesage** [comment]
> the vast majority of them will never use git or diff.

Absolutely. BUT if you offer them an easy way using just a click, they'll adopt it. It's all about not scaring them — especially college students are very open to new cool magic as I can say from my own experience, but of course they need to have some guidance.

My current vision is: Right click a folder "make git" which'll simply run git init in the background. Then right click that one, say "create snapshot" which'll simply run git add and git commit in the background. And then right click "Available snapshots ..." which'll simply list all commits, and click on one, say "revert" and it'll basically pull that commit.

Something like that; should be easy enough for non-technical folks, and still work like a charm!

**Comment by rsgranne** [comment]
Oh boy. I respect your intelligence, skills, & work, but you & I have very different understandings of what "normal" people want to learn & are willing to learn!

Anyway, I’d find CriticMarkup useful in Zettlr. If no one else would, or you’re not interested in implementing it, that’s cool.

**Comment by jochmann** [comment]
I am currently scouring the interwebs to find a windows-native app that offers critic markup support - precisely because I have a user I need to onboard into a markdown based collaborative workflow and using git is a nonstarter. Take this as a +1 (and a call for help if you do know a current markdown editor that allows previewing critic markup). Cheers!

**Comment by rsgranne** [comment]
On my Mac, nvUltra has built-in support. 

The best Windows app for Markdown that I’ve seen is [Markdown Monster](https://markdownmonster.west-wind.com/), but alas, I don’t see any mention of CriticMarkup support. It’s open source, however, & actively developed, so maybe the maintainers might be open to adding support. 

I see there’s a cross-platform tool called [Marko Editor](http://marko-editor.com/marko-editor/) that support CriticMarkup. A good place to look for more is on the [CriticMarkup Support Forum](https://criticmarkup.tenderapp.com/help).

**Comment by jochmann** [comment]
interesting to learn that you're willing to move away from nvUltra to zettlr. I'm still stuck in nvALT, looking at the closed beta. When I was looking for a Windows app your issue was the only lead towards a critic markup feature in a Windows based Markdown-editor. Marko is offline, btw (probably abandoned, too). I've even installed Software from Downloads on the wayback machine and abandoned atom-packages. There really is no current, actively developed critic markup support in Windows :(

Re: Discussion - The inline-comment feature alone sets it far apart from a diff in my opinion. I've used critic markup together with GitHub. For me, pne is for collaborating on text, the other for version control.

**Comment by rsgranne** [comment]
No, I’m probably going to stick with nvUltra for now, but I’m going to keep an eye on Zettlr, as it’s very impressive. As for nvUltra, you’ll be happy to know that is far better than nvAlt.

**Comment by guerrajose** [comment]
First thank you for Zettlr, an incredible software.

We are a small consultancy and some of us use Zettlr on a personal manner to draft text. Yet, we still have to revert to Word for collaboration among us: mainly to use track changes and comments.

I didn't try criticmarkup, but I spent quite some time looking for ways on how to collaborate with markdown, and I ended up there. And then on this issue.

We tried https://draftin.com/ and the collaboration functionality worked pretty well. But the interface was not that practical and we like to use free and open source software as much as possible.

Regarding the use of Git, we don't find it appropriate for "collaboration" on text for one main reason (as far as I understand git):

* We can't accept/reject/comment specific sections of a text.
* Let's say we have a document we review and we modify text at 50 locations. With Git, we can only "compare" different versions of the whole text, and thus we can't "cherry-pick" the changes we want to keep or to discard, nor can we comment each of them separately.
* I am sure the fact that we got used to how "track-changes" work in Word has modified our expectations, but the fact is that it is most useful to be able to see all changes and comments of a document, to cherry-pick what we accept or reject, and to be able to add new edits in "trackchanges".

I don't know if criticmarkup is the solution, but adding some collaboration features on Zettlr would be incredible (but I imagine pretty difficult).

**Comment by conig** [comment]
Resurrecting this issue as this feature would be really useful. Collaborating in markdown is a pain currently.
An accept/reject suggestion interface + CriticMarkup syntax would be revolustionary IMO.

**Comment by salix-d** [comment]
I'm currently trying different filters so I can keep working in zettlr even if the people reviewing my text use docx to keep track of change... 
The CriticMarkup seems a lot lesss complicated was hoping the support had already been added :/

**Comment by the-solipsist** [comment]
I think it might help if someone actually worked out a workflow for usage of CriticMarkup, especially when working with those using things like track changes in docx files. 

Going through the discussion at [this Pandoc thread](https://github.com/jgm/pandoc/issues/5430) and looking at [PanDiff](https://github.com/davidar/pandiff) and [PanCritic](​https://github.com/ickc/pancritic) might help get some ideas flowing, including thinking through some potential obstacles.

**Comment by frederik-elwert** [comment]
I would also welcome support for Critic Markup in Zettlr! Reviewing some else’s texts is really not solved well in plain text editing at the moment.

Sublime’s [MarkdownEditing](https://sublimetext-markdown.github.io/MarkdownEditing/#critic-markup) package has quite elaborate support for Critic Markup, once you know the keyboard shortcuts. E.g., you can select something and mark it as deletion via `Alt+C Alt+D`, and then someone else can accept that deletion by just moving the cursor inside the deleted segment and hit `Alt+Enter` (or discard it by hitting `Alt+Backspace`).

For Zettlr, I guess something more obvious would be nice. Here’s my first thoughts: For reviewing documents, a simple "track changes" button to toggle critic mode would be enough. Once in critic mode, selecting and deleting text would actually insert the corresponding markup, same for typing (the new text would be wrapped in addition tags), or selecting and typing (which would insert substitution tags). Highlights and comments might need their own buttons (although they also work well by just using the corresponding markup, and Zettlr even has partial support for highlights already).

For accepting/rejecting changes, a different UI might be required, e.g. accept and reject buttons that become visible when the cursor is inside a critic markup span.

**Comment by nathanlesage** [comment]
Quick update on this front: A few months ago I started to write a parser for critic markup that can appropriately parse the syntax on top of a Markdown parser: https://github.com/nathanlesage/lang-criticmarkup/

The only big amount of work would now involve a custom editing mode to "track changes".
